security/pkg/pki/util/keycertbundle.go

```diff
@@ -318,18 +318,12 @@ func Verify(certBytes, privKeyBytes, certChainBytes, rootCertBytes []byte) error
 	return nil
 }
 
-func extractCertExpiryTimestamp(certType string, certPem []byte) (float64, error) {
+func extractCertExpiryTimestamp(certType string, certPem []byte) (*time.Time, error) {
 	cert, err := ParsePemEncodedCertificate(certPem)
 	if err != nil {
-		return -1, fmt.Errorf("failed to parse the %s: %v", certType, err)
-	}
-
-	end := cert.NotAfter
-	expiryTimestamp := float64(end.Unix())
-	if end.Before(time.Now()) {
-		return expiryTimestamp, fmt.Errorf("expired %s found, x509.NotAfter %v, please transit your %s", certType, end, certType)
+		return nil, fmt.Errorf("failed to parse the %s: %v", certType, err)
```

Comment:
Unsure about this change.
I feel like this function should leave expiration checking to the caller, or mention expiration checking to the caller.
Regardless, I don't think using -1 as a sentinel value for an error value makes much sense since -1 would indicate 1 second before new year 1970.

Reply:
I tend to appreciate adhering to the unix philosophy of "doing one thing and doing it well" but I wonder if this expiration checking is too complicated to expect the caller to do. Seems safer in this case to just change the name or isolate the change itself in a separate PR and ensure the right tests have been created/modified.

It seems like the goal of this function is to get the `expiryTimestamp` which it seems to _always_ do. Callers should check for `err != nil` regardless, it's just whether or not it makes sense to have them do the float64 conversion every time as well.

Reply:
I think this is a valid change, where returning `nil` for the first return value is a better indicator that there was an error. Regardless, I agree most callers will check `err -= nil`.

Reply:
I think leaving it in this PR is better because it affects the error handling logic for recording metrics.
Regardless, the usage of this function strongly suggests that it should **not** return an error if the cert has expired.
For example, in `keycertbundle.go:269`, we have (in master)

```go
// TimeBeforeCertExpires returns the time duration before the cert gets expired.
// It returns an error if it failed to extract the cert expiration timestamp.
// The returned time duration could be a negative value indicating the cert has already expired.
func TimeBeforeCertExpires(certBytes []byte, now time.Time) (time.Duration, error) {
	if len(certBytes) == 0 {
		return 0, fmt.Errorf("no certificate found")
	}

	certExpiryTimestamp, err := extractCertExpiryTimestamp("cert", certBytes)
	if err != nil {
		return 0, fmt.Errorf("failed to extract cert expiration timestamp: %v", err)
	}

	certExpiry := certExpiryTimestamp.Sub(now)
	return certExpiry, nil
}
```

If `extractCertExpiryTimestamp` returned an error on expired certs, then this function would return 0, not a negative value as the documentation suggests.

Further, the only use of `TimeBeforeCertExpires` (other than testing) is in `ca.go:437` which explicitly checks for a negative expiration time given that there is no error, which implies `extractCertExpiryTimestamp` should not error if the timestamp has expired.

Finally, the only other uses of `extractCertExpiryTimestamp` are in `ExtractRootCertExpiryTimestamp` and `ExtractCACertExpiryTimestamp` both of which are (only) used for metrics.

Reply:
I think the normal pattern is to return time.Time(0) - there is check for isZero(). Returning a pointer doesn't seem right.

Reply:
Not much diff for me time.Time or its pointer, from both we can know the  expire time.